src/v/iceberg/schema_protobuf.cc

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/schema_protobuf.h"
+
+#include "iceberg/conversion_outcome.h"
+#include "iceberg/datatypes.h"
+
+#include <seastar/core/sstring.hh>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <google/protobuf/descriptor.h>
+#include <google/protobuf/descriptor.pb.h>
+
+#include <deque>
+
+namespace iceberg {
+
+namespace {
+
+namespace pb = google::protobuf;
+
+/**
+ * Schema conversion from protocol buffers to internal Redpanda iceberg schema
+ */
+using field_outcome = conversion_outcome<nested_field_ptr>;
+using struct_outcome = conversion_outcome<struct_type>;
+using type_stack = std::deque<const pb::Descriptor*>;
+static constexpr int max_recursion_depth = 64;
```

Comment:
Technically the spec allows 100: https://protobuf.dev/programming-guides/proto-limits/#depth, but that includes recursion, so maybe it's OK to have lower? 100 hopefully is safe too. Do we have tests for this bit?